controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/resource/ResourceSnapshot.java

```diff
@@ -41,37 +44,26 @@ public static ResourceSnapshot from(List<Node> nodes, Instant timestamp, ZoneId
                                                  .map(node -> node.owner().get())
                                                  .collect(Collectors.toSet());
 
+        Set<Version> versions = nodes.stream()
+                .map(Node::currentVersion)
+                .collect(Collectors.toSet());
+
         if (applicationIds.size() != 1) throw new IllegalArgumentException("List of nodes can only represent one application");
+        if (versions.size() != 1) throw new IllegalArgumentException("List of nodes can only represent one version");
```

Comment:
This constraint is correct, but the node list needs to be grouped by version beforehand. Otherwise the maintainer will fail whenever an application is upgrading, and nodes in the same cluster are on different Vespa versions.

Reply:
Good catch. I created a grouping on major version in addition to architecture.